### PR TITLE
Simplify compiler dependent checks

### LINKED ISSUE #2783: [FEA]: Define compiler id macros to true value
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

General CCCL

### Is your feature request related to a problem? Please describe.

CCCL supports various host compilers. Currently, the host compiler dependent code is implemented as:
```cpp
#if defined(_CCCL_COMPILER_MYCOMPILER)
// ...
#endif // defined(_CCCL_COMPILER_MYCOMPILER)
```
This is a simple case, however, if the code is dependent on multiple host compilers, the conditions start to get very verbose.

### Describe the solution you'd like

We could take an advantage of the fact that when preprocessor `#if` statement is evaluated, unknown identifiers are replaced with `0`. If define compiler id macros to `1`, the `defined(_CCCL_COMPILER_MYCOMPILER)` can be replaced with simple `_CCCL_COMPILER_MYCOMPILER`. It makes the code more readable.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by davebayer**
## Description

Closes #2783.

Compiler id macros are now expanded to `1`. It allows replacing `defined(arg)`  with simple `arg` in preprocessor if statements.

## Checklist
- [x] I am familiar with the [Contributing Guidelines]().
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by miscco** [comment]
Lets discuss this with the other maintainers

**Comment by miscco** [comment]
@davebayer I have reopened this.

My general concerns were about making any of these macros public and also spending limited maintainer time on it.

With you having put in the time I am happy to revisit that

**Review by bernhardmgruber — APPROVED** [review]
Alright. I like the refactoring. Thanks a lot for doing the work!

**Comment by miscco** [comment]
@bernhardmgruber and I have been talking a bit internally and we come around liking this a lot.

We were wondering whether we could directly set the macro to the compiler version. 

However that would not work immediately because `#if _CCCL_COMPILER_CLANG < 70000` would be true for all compilers except clang

**Comment by bernhardmgruber** [comment]
> We were wondering whether we could directly set the macro to the compiler version.

But please propose a separate PR for that! Thx a lot!

File: cudax/include/cuda/experimental/__container/uninitialized_async_buffer.cuh [hunk 1]
```diff
@@ -34,8 +34,7 @@
 
 #include <cuda/experimental/__memory_resource/any_resource.cuh>
 
-#if _CCCL_STD_VER >= 2014 && !defined(_CCCL_COMPILER_MSVC_2017) \
-  && defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#if _CCCL_STD_VER >= 2014 && !_CCCL_COMPILER_MSVC_2017 && defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
```
[the remark this answers is not in the document]
Comment (davebayer) [reply cudax/include/cuda/experimental/__container/uninitialized_async_buffer.cuh:37]: What inconsistency do you have in mind? That some macros (like `LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE`) are checked by definition and others (like `_CCCL_STD_VER`) by value? Maybe all internal macros could copy this approach and be defined to `1` or be undefined in the future.

[the remark this answers is not in the document]
Comment (miscco) [reply cudax/include/cuda/experimental/__container/uninitialized_async_buffer.cuh:37]: I would not want to change any other macros

This PR is already massive as is and user facing macros are a totally different story

**Comment by davebayer** [comment]
pre-commit.ci autofix

**Comment by ericniebler** [comment]
this idiom is worth considering: https://www.fluentcpp.com/2019/05/28/better-macros-better-flags/

i have mostly switched to function-like macros in my own code. it would also permit things like `#if _CCCL_COMPILER(MSVC)` which would hard-error if the compiler name is misspelled:

```c++
#define EXPAND(...) __VA_ARGS__
#define TAIL(X, ...) __VA_ARGS__

#ifdef __clang__
#  define COMPILER_CLANG (~, 1)
#elif __GNUC__
#  define COMPILER_GCC (~, 1)
#elif _MSC_VER
#  define COMPILER_MSVC (~, 1)
#endif

#ifndef COMPILER_CLANG
#  define COMPILER_CLANG (~, 0)
#endif
#ifndef COMPILER_GCC
#  define COMPILER_GCC (~, 0)
#endif
#ifndef COMPILER_MSVC
#  define COMPILER_MSVC (~, 0)
#endif

#define COMPILER(X) EXPAND(TAIL COMPILER_ ## X)

#if COMPILER(CLANG)
#  error CLANG!
#endif

#if COMPILER(GCC)
#  error GCC!
#endif

#if COMPILER(MSVC)
#  error MSVC!
#endif

#if COMPILER(OOPS) // hard error here
#  error should never get here
#endif
```

https://godbolt.org/z/EYfGEMoYd

**Comment by davebayer** [comment]
> the trouble here is that someone can still do `#ifdef _CCCL_COMPILER_MSVC`, and it would do the wrong things. this other idiom is worth considering: https://www.fluentcpp.com/2019/05/28/better-macros-better-flags/
> 
> i have mostly switched to function-like macros in my own code. it would also permit things like `#if _CCCL_COMPILER(MSVC)` which would hard-error if the compiler name is misspelled:
> 
> ```c++
> #define TAIL(X, ...) __VA_ARGS__
> #define EVAL(X, ...) X(__VA_ARGS__)
> 
> #ifdef __clang__
> #  define COMPILER_CLANG ~, 1
> #elif __GNUC__
> #  define COMPILER_GCC ~, 1
> #elif _MSC_VER
> #  define COMPILER_MSVC ~, 1
> #endif
> 
> #ifndef COMPILER_CLANG
> #  define COMPILER_CLANG ~, 0
> #endif
> #ifndef COMPILER_GCC
> #  define COMPILER_GCC ~, 0
> #endif
> #ifndef COMPILER_MSVC
> #  define COMPILER_MSVC ~, 0
> #endif
> 
> #define COMPILER(X) EVAL(TAIL, COMPILER_ ## X)
> 
> #if COMPILER(CLANG)
> #  error CLANG!
> #endif
> 
> #if COMPILER(GCC)
> #  error GCC!
> #endif
> 
> #if COMPILER(MSVC)
> #  error MSVC!
> #endif
> 
> #if COMPILER(OOPS) // hard error here
> #  error should never get here
> #endif
> ```
> 
> https://godbolt.org/z/75Ge57PKe

I was also considering similar approach, but, I felt like it overcomplicates thins too much. However, if this is an approach you consider, I would suggest a similar approach to checking the compiler versions. Instead of
```cpp
#if _CCCL_COMPILER_GCC && _CCCL_GCC_VERSION >= 100000
// do stuff
#endif
```
we could do something like
```cpp
#define _CCCL_COMPILER_VERSION_GTE(comp, major, minor, patch) (_CCCL_COMPILER(comp) && _CCCL_COMPILER_VERSION(comp) >= major##minor##patch)

#if _CCCL_COMPILER_VERSION_GTE(GCC, 10, 0, 0)
// do stuff
#endif
```

**Comment by ericniebler** [comment]
> I was also considering similar approach, but, I felt like it overcomplicates thins too much

i'm of a different opinion. whenever i use one of the CCCL compiler/version checking macros, i need to double-check that i'm using them correctly. i'm forgetful and paranoid. the function-like macros would relieve my paranoia.

> I would suggest a similar approach to checking the compiler versions

how about something like this:

```c++
#if _CCCL_COMPILER(MSVC [, (EQ|NEQ|LT|LTE|GT|GTE), \(major [, minor [, patch]]\)])
```

so you can do:

```c++
#if _CCCL_COMPILER(MSVC)
```

or

```c++
#if _CCCL_COMPILER(MSVC, LT, (19,36))
```

File: cudax/include/cuda/experimental/__container/uninitialized_async_buffer.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericniebler) [reply cudax/include/cuda/experimental/__container/uninitialized_async_buffer.cuh:37]: we could always put this in cccl's config header:

```c++
#if defined(LIBCUDACXX_OLD_PUBLIC_MACRO)
#  define CCCL_NEW_PUBLIC_MACRO() 1
#else
#  define CCCL_NEW_PUBLIC_MACRO() 0
#endif
```

**Comment by davebayer** [comment]
> > I was also considering similar approach, but, I felt like it overcomplicates thins too much
> 
> i'm of a different opinion. whenever i use one of the CCCL compiler/version checking macros, i need to double-check that i'm using them correctly. i'm forgetful and paranoid. the function-like macros would relieve my paranoia.
> 
> > I would suggest a similar approach to checking the compiler versions
> 
> how about something like this:
> 
> ```c++
> #if _CCCL_COMPILER(MSVC [, (EQ|NEQ|LT|LTE|GT|GTE), \(major [, minor [, patch]]\)])
> ```
> 
> so you can do:
> 
> ```c++
> #if _CCCL_COMPILER(MSVC)
> ```
> 
> or
> 
> ```c++
> #if _CCCL_COMPILER(MSVC, LT, (19,36))
> ```

What about something like this (see in [Compiler Explorer](https://godbolt.org/z/1EMen61Ec):
```cpp
#define _CCCL_COMPILER_GCC     1
#define _CCCL_COMPILER_CLANG   2
#define _CCCL_COMPILER_MSVC    3
#define _CCCL_COMPILER_NVHPC   4
#define _CCCL_COMPILER_NVRTC   5
#define _CCCL_COMPILER_ICC     6
#define _CCCL_COMPILER_UNKNOWN 7

#define _CCCL_COMPILER_MAKE_VERSION_GCC(major, minor, patch)   (major * 10000 + minor * 100 + patch)
#define _CCCL_COMPILER_MAKE_VERSION_CLANG(major, minor, patch) (major * 10000 + minor * 100 + patch)
#define _CCCL_COMPILER_MAKE_VERSION_MSVC(major, minor, build)  (major * 1000000 + minor * 100000 + build)
#define _CCCL_COMPILER_MAKE_VERSION_NVHPC(major, minor, patch) (major * 10000 + minor * 100 + patch)

#if defined(__INTEL_COMPILER)
#  define _CCCL_COMPILER _CCCL_COMPILER_ICC
#elif defined(__NVCOMPILER)
#  define _CCCL_COMPILER               _CCCL_COMPILER_NVHPC
#  define _CCCL_COMPILER_VERSION_NVHPC _CCCL_COMPILER_MAKE_VERSION_NVHPC(__NVCOMPILER_MAJOR__, __NVCOMPILER_MINOR__, __NVCOMPILER_PATCHLEVEL__)
#elif defined(__clang__)
#  define _CCCL_COMPILER               _CCCL_COMPILER_CLANG
#  define _CCCL_COMPILER_VERSION_CLANG _CCCL_COMPILER_MAKE_VERSION_CLANG(__clang_major__, __clang_minor__, __clang_patchlevel__)
#elif defined(__GNUC__)
#  define _CCCL_COMPILER             _CCCL_COMPILER_GCC
#  define _CCCL_COMPILER_VERSION_GCC _CCCL_COMPILER_MAKE_VERSION_GCC(__GNUC__, __GNUC_MINOR__, __GNUC_PATCHLEVEL__)
#elif defined(_MSC_VER)
#  define _CCCL_COMPILER              _CCCL_COMPILER_MSVC
#  define _CCCL_COMPILER_VERSION_MSVC _MSC_FULL_VER
#elif defined(__CUDACC_RTC__)
#  define _CCCL_COMPILER _CCCL_COMPILER_NVRTC
#else
#  define _CCCL_COMPILER _CCCL_COMPILER_UNKNOWN
#endif

#define _CCCL_MSVC_2017 192000000
#define _CCCL_MSVC_2019 193000000
#define _CCCL_MSVC_2022 194000000

#define _CCCL_OPERATOR_EQ  ==
#define _CCCL_OPERATOR_NEQ !=
#define _CCCL_OPERATOR_LT  <
#define _CCCL_OPERATOR_LTE <=
#define _CCCL_OPERATOR_GT  >
#define _CCCL_OPERATOR_GTE >=

#define _CCCL_IF_COMPILER(comp)                  (_CCCL_COMPILER == _CCCL_COMPILER_##comp)
#define _CCCL_IF_COMPILER_VERSION(comp, op, ...) (_CCCL_IF_COMPILER(comp) && (_CCCL_COMPILER_VERSION_##comp _CCCL_OPERATOR_##op _CCCL_COMPILER_MAKE_VERSION_##comp(__VA_ARGS__)))

#if _CCCL_IF_COMPILER(GCC)
# warning "GCC"
#elif _CCCL_IF_COMPILER(CLANG)
# warning "CLANG"
#elif _CCCL_IF_COMPILER(MSVC)
# pragma message("MSVC")
#elif _CCCL_IF_COMPILER(NVHPC)
# warning "NVHPC"
#elif _CCCL_IF_COMPILER(NVRTC)
# warning "NVRTC"
#endif

#if _CCCL_IF_COMPILER_VERSION(GCC, LT, 15, 0, 0)
# warning "GCC less than 15"
#elif _CCCL_IF_COMPILER_VERSION(CLANG, GT, 15, 0, 0)
# warning "CLANG greater than 15"
#elif _CCCL_IF_COMPILER_VERSION(MSVC, GT, 15, 0, 0)
# pragma message("MSVC greater than 15")
#elif _CCCL_IF_COMPILER_VERSION(NVHPC, NEQ, 1, 0, 0)
# warning "NVHPC not equal to 1"
#elif _CCCL_IF_COMPILER(NVRTC)
# warning "NVRTC"
#endif

#if _CCCL_IF_COMPILER_VERSION(MSVC, GT, _CCCL_MSVC_2017)
# pragma message("MSVC 2017 and newer")
#endif
```
It does not work yet for MSVC, but you get the idea. Nice thing about this solution is that version checking can be done separately for each compiler, e. g. we can use only minor and major versions for MSVC. We can also simplify all `_CCCL_CUDACC_BELOW_*_*` checks.

**Comment by ericniebler** [comment]
this is very nice. i would make a very small tweak to the syntax:

```c++
#if _CCCL_IF_COMPILER(GCC)
```

and

```c++
#if _CCCL_IF_COMPILER(GCC, LT(15, 0, 0))
```

https://godbolt.org/z/TG53sKcf8

we will also have to uglify everything, like `_CCCL_IF_COMPILER(_GCC, _LT(15, 0, 0))`.

**Comment by davebayer** [comment]
> this is very nice. i would make a small tweak to the syntax, which you are free to take or not:
> 
> ```c++
> #if _CCCL_IF_COMPILER(GCC)
> ```
> 
> and
> 
> ```c++
> #if _CCCL_IF_COMPILER(GCC, LT(15, 0, 0))
> ```
> 
> https://godbolt.org/z/TG53sKcf8
> 
> we will also have to uglify everything, like `_CCCL_IF_COMPILER(_GCC, _LT(15, 0, 0))`.
> 
> EDIT: oh, and as currently written, typos in the compiler name are not being caught:
> 
> ```c++
> #if _CCCL_IF_COMPILER(WHOOPS) // should be a hard error
> ```
> 
> one last thing ... since this is generally going to appear after a `#if` or `#elif`, it seems redundant to have `_IF_` in the macro name. maybe call it `_CCCL_COMPILER_IS` or just `_CCCL_COMPILER`.

Wow, this is a very nice solution. But there is a big problem. If a user does this:
```cpp
#define GCC CLANG
#define MSVC
```
the code does not work as intended. The advantage of my solution is that it cannot be influenced by external macros.

See [yours](https://godbolt.org/z/Yax6rsjoo) vs [mine](https://godbolt.org/z/dv64rqbnr).

PS: Just the `_CCCL_IF_COMPILER(...)` inside `_CCCL_IF_COMPILER_VERSION(...)` had to be replaced with its contents.

**Comment by miscco** [comment]
I am not a fan of `_CCCL_COMPILER_IS(` compared to just `_CCCL_COMPILER` but nothing I would die for

What I definitely do not like is `_CCCL_COMPILER_IS` and `_CCCL_COMPILER_VERSION_IS` those should just be one macro

**Comment by miscco** [comment]
To be more precise what I have in mind is something along those lines
https://godbolt.org/z/dGYEEacE7

**Comment by davebayer** [comment]
I've started to rework the compiler checking, but this is even bigger rework than before. I think it would be better to implement this compiler by compiler in separate PRs.

**Comment by miscco** [comment]
Yeah I believe we can start with a single compiler and only define that to a value and leave the rest as is.

**Comment by miscco** [comment]
It would be best to use one like NVHPC / ICC / NVRTC were we rarely use any version checks

**Pull request closed without merging**